### ISSUE TITLE
PCA9635 delay between write1(..) commands and configuration clarity 

### ISSUE THREAD

**Issue by ArtimusSteven** [RobTillaart/PCA9635#24]
Hello, 
I have been utilizing the PCA9635 chip C++ files in order to independently control 16 LEDs (the total available amount on chip) within my circuit system. I have a Teensy 3.6 as the master and 2 PCA9635s (32 LEDs total) as the slaves. 

When using version 0.3.3 (before realizing most recent version) the output pwm signals successuflly reached the 3.3 supply voltage of the teensy 3.6; but there is a delay of ~270 micro seconds in between consecutive write1(...) commands. I am currently trying to determine if this is an intrinsic delay of the hardware or a config problem.

Below is the code i use with version 0.3.3. The setOutputMode method was slightly altered in order to successfully invert the pin outputs (this was fixed with more recent versions). 

```cpp
PCA9635 ledArray(0xF);
PCA9635 ledArray2(0x1F);

//pin assignments etc etc not showing 


setup(){

  ledArray.begin();
  ledArray2.begin();
  ledArray.setOutputMode(invert); //see updated method below 
  ledArray2.setOutputMode(invert);

/////////updated version of setoutputmode in PCA9635.cpp file (version 0.3.3)///////////////

oid PCA9635::setOutputMode(bool invert) {
  uint8_t oldmode = readReg(PCA9635_MODE2);
  uint8_t newmode;
  if (invert) {
    newmode = oldmode | PCA9635_INVERT;
  } else {
    newmode = oldmode & ~PCA9635_INVERT;
  }
  writeReg(PCA9635_MODE2, newmode);


////////////////////////////////////////////////////////////////////////////////////////////////////////////////

  for (int ch = 0; ch < 16; ch++)
  {
    ledArray.setLedDriverMode(ch, PCA9635_LEDPWM);
    ledArray2.setLedDriverMode(ch, PCA9635_LEDPWM);
    ledArray.write1(ch, 0);
    ledArray2.write1(ch, 0);
  }

}


loop(){
 curr = micros();

  if (sw1 && (curr - prev1) >= OnTime1)
  {
    ledArray2.write1(H15, 0);
    ledArray2.write1(H16, 255);
    sw1 = !sw1;
    prev1 = curr;
  }

  if ((curr - prev1) >= OffTime1)
  {
    ledArray2.write1(H16, 0);
    ledArray2.write1(H15, 255);

    sw1 = !sw1;
    prev1 = curr;
  }

}
```

I have recently implemented a similar script with the updated version 0.4.4. Below is the config code with the new version. My problem with this script is that the output pins H15 and H16 are only reaching mV values instead of the desired 3.3V supplied by the teensy 3.6.

Code:

```cpp
PCA9635 ledArray(0xF);
PCA9635 ledArray2(0x1F);

// pin assignments etc etc 


setup(){

  ledArray.begin();
  ledArray2.begin();

  uint8_t prev11 = ledArray.getMode1();
  uint8_t prev12 = ledArray.getMode2();
  uint8_t prev21 = ledArray2.getMode1();
  uint8_t prev22 = ledArray2.getMode2();
  
  uint8_t mode11_maskc = prev11 | PCA9635_MODE1_NONE;
  uint8_t mode12_maskc = prev12 | PCA9635_MODE2_INVERT;  //Invert is desired so that pin outputs are low on startup

  uint8_t mode21_maskc = prev21 | PCA9635_MODE1_NONE;
  uint8_t mode22_maskc = prev22 | PCA9635_MODE2_INVERT;


  ledArray.setMode2(mode12_maskc);
  ledArray2.setMode2(mode22_maskc);


  for (int ch = 0; ch < 16; ch++)
  {
    ledArray.setLedDriverMode(ch, PCA9635_LEDPWM);
    ledArray2.setLedDriverMode(ch, PCA9635_LEDPWM);
    ledArray.write1(ch, 0);
    ledArray2.write1(ch, 0);
  }

}

loop(){

 curr = micros();

  if (sw1 && (curr - prev1) >= OnTime1)
  {
    ledArray2.write1(H15, 0);
    ledArray2.write1(H16, 255);
    sw1 = !sw1;
    prev1 = curr;
  }

  if ((curr - prev1) >= OffTime1)
  {
    ledArray2.write1(H16, 0);
    ledArray2.write1(H15, 255);

    sw1 = !sw1;
    prev1 = curr;
  }

}
```

Please reach back out if you need further clarity. Any help would be greatly appreciated.

Best,

Artimus Steven 

(updated code tags for readability)

**Comment by RobTillaart** [comment]
Thanks for the issue,

Had a quick look at the code and I noticed is that 
```cpp
if ((curr - prev1) >= OffTime1)
```
seems to miss the test for ```!sw1```.

Can you check?

I need to check if I have free hardware around to make a test setup.

**Comment by RobTillaart** [comment]
I would write the loop as follows as this way:
- loop() does minimal testing per loop and 
- the off-code cannot interfere with the on-code.

```cpp
loop()
{
  curr = micros();
  if (sw1)
  {
    if ((curr - prev1) >= OnTime1)
    {
      ledArray2.write1(H15, 0);
      ledArray2.write1(H16, 255);
      sw1 = !sw1;
      prev1 = curr;
    }
  }
  else
  {
    if ((curr - prev1) >= OffTime1)
    {
      ledArray2.write1(H16, 0);
      ledArray2.write1(H15, 255);
      sw1 = !sw1;
      prev1 = curr;
    }
  }
}
```

**Comment by RobTillaart** [comment]
a minor issue here, you write

```cpp
  uint8_t mode11_maskc = prev11 | PCA9635_MODE1_NONE;
  uint8_t mode12_maskc = prev12 | PCA9635_MODE2_INVERT;  //Invert is desired so that pin outputs are low on startup

  uint8_t mode21_maskc = prev21 | PCA9635_MODE1_NONE;
  uint8_t mode22_maskc = prev22 | PCA9635_MODE2_INVERT;
```

**PCA9635_MODE1_NONE** == 0x00 so the **OR** doesn't do anything

You probably meant (please verify)
```cpp
  uint8_t mode11_maskc = PCA9635_MODE1_NONE;  //  clear all mode1 flags
  uint8_t mode12_maskc = prev12 | PCA9635_MODE2_INVERT;   // set the INVERT mode

  uint8_t mode21_maskc = PCA9635_MODE1_NONE;
  uint8_t mode22_maskc = prev22 | PCA9635_MODE2_INVERT;
```

----

Furthermore
```cpp
  ledArray.setMode2(mode12_maskc);
  ledArray2.setMode2(mode22_maskc);
```
 You do not call **setMode1()** while you prepare the mask?

----

Note to myself: consider to add **clearMode1()** and **clearMode2()** functions.

**Comment by RobTillaart** [comment]
BTW- no PCA9635 around here for testing.

**Comment by RobTillaart** [comment]
Can you run this test sketch and post the output to see timing of **write1()**?

```cpp
//    FILE: PCA9635_performance.ino
//  AUTHOR: Rob Tillaart
// PURPOSE: test PCA9635 library

#include "Arduino.h"
#include "Wire.h"
#include "PCA9635.h"

PCA9635 ledArray(0x20);

uint32_t start, stop;

void setup()
{
  Serial.begin(115200);
  Serial.print("PCA9635 LIB version: ");
  Serial.println(PCA9635_LIB_VERSION);
  Serial.println();

  ledArray.begin();

  delay(100);  //  flush all Serial interrupts
  start = micros();
  ledArray.write1(0, 127);  //  ~50%
  stop = micros();
  Serial.print("ONE:\t");
  Serial.println(stop - start);

  delay(100);  //  flush all Serial interrupts
  start = micros();
  ledArray.write1(0, 127);  //  ~50%
  ledArray.write1(1, 128);  //  ~50%
  stop = micros();
  Serial.print("TWO:\t");
  Serial.println(stop - start);

  delay(100);  //  flush all Serial interrupts
  uint8_t n = 123;
  start = micros();
  ledArray.writeN(0, &n, 1);  //  ~50%
  stop = micros();
  Serial.print("wrN:\t");
  Serial.println(stop - start);

  Serial.println("done...");
}

void loop()
{
}

// -- END OF FILE --
```

Indicate timing of one call and two calls and should give an indication of the costs of wrapping **writeN()**.
See if there is some serious delay on the calling side.

**Comment by RobTillaart** [comment]
Is your goal to use **PWM** or just **ON/OFF** ?

**Comment by RobTillaart** [comment]
From the start of the datasheet

_Each LED output has its own 8-bit resolution (256 steps) fixed frequency individual PWM controller that operates at 97 kHz_

#### hypothesis
It could be that when writing a new value to a channel it will only be updated when a new PWM cycle starts.

However  97 kHz in 256 steps imply a ~380 Hz refresh frequency. 
On average one should see a delay of 1000000/380/2 = ~1300 us. which is much more than the observed 270 us.

So does not explain, but might be related. 
to be continued.

**Comment by RobTillaart** [comment]
just a thought:
You can use the OE (output Enable) line to have all leds OFF at start without need to INVERT.

**Comment by RobTillaart** [comment]
#### par 7.3.1 about Sleep bit

_[1] It takes 500 us max. for the oscillator to be up and running once SLEEP bit has been set to logic 0. Timings on LEDn outputs are not guaranteed if PWMx, GRPPWM or GRPFREQ registers are accessed within the 500 us window._

Might be a cause for unexpected delay given that you did not set mode1  (see above).

> You do not call setMode1() while you prepare the mask?

**Comment by RobTillaart** [comment]
Went through the datasheet again, except from the **sleep** above no clues about unexpected delays in the order of ~250 us.

**Comment by ArtimusSteven** [comment]
I appreciate the feedback and next steps proposed. I am just seeing these. Will implement and relay results shortly.

**Comment by ArtimusSteven** [comment]
Output for write1 timing script: 
PCA9635 LIB version: 0.4.4

ONE:	98
TWO:	205
wrN:	98
done...

**Comment by ArtimusSteven** [comment]
The goal is to use PWM as opposed to ON/OFF. The 0-255 variability is essential to the system.

**Comment by ArtimusSteven** [comment]
how did you determine this delay: On average one should see a delay of 1000000/380/2 = ~1300 us

**Comment by ArtimusSteven** [comment]
![pca9635_pwm_data](https://user-images.githubusercontent.com/118404691/230650929-d673b832-d486-4b24-8596-2ba3944812aa.png)

This is a link to a png image taken from the oscilliscope i am using. The issue here is the amplitude of the pwm signal not reaching the desired supply voltage of the teensy 3.6 (3.3V). It appears that the Vp-p amplitude is closer to 3.1mV. Is there something you are aware of that would constrain the output voltage in this way?

**Comment by RobTillaart** [comment]
> how did you determine this delay: On average one should see a delay of 1000000/380/2 = ~1300 us

The datasheet states the PWM uses a frequency of 97 KHz.  So that is the freq of the individual pwm steps. These are rounded 10 us, so the PWM goes in 0..255 steps of ~10 us. Is after a max of ~2550 us the next PWM pulse starts over again. 

As on average the register changes in the middle of such PWM cycle, it takes about half the cycle time before it becomes active. 2550/ 2 = ~1300 us.

That is the hypothesis.

**Comment by RobTillaart** [comment]
> Output for write1 timing script: PCA9635 LIB version: 0.4.4
> 
> ONE: 98 TWO: 205 wrN: 98 done...

Ok wrapping costs no extra time, and there seems no extra blocking time with multiple calls short after one another.

**Comment by RobTillaart** [comment]
.> It appears that the Vp-p amplitude is closer to 3.1mV. Is there something you are aware of that would constrain the output voltage in this way?

Can you pull the OE pin LOW (GND) ? 
From my head (no datasheet on my phone) this line should be LOW.

**Comment by ArtimusSteven** [comment]
Yes, the OE pin is pulled down to ground via a 10k ohm resistor

**Comment by ArtimusSteven** [comment]
PCA9635 ledArray(0xF);
PCA9635 ledArray2(0x1F); //I used an I2C scanner to detect which addresses were found 

Do these initial I2C addresses need to end in a 0 if I want to use them for writes? They both currently end (far right bit) with 1.

**Comment by RobTillaart** [comment]
I2C addresses are typical 7 bits 0..127  (in fact several sre reserved).
Depending on read/write a 0 or 1 is added. 
Do these addresses seem ok

**Comment by ArtimusSteven** [comment]
I will experiment more with slave addresses and mode configuration bits and report back

**Comment by ArtimusSteven** [comment]
What are the procedures to configure control registers for each individual PWM address (as opposed to configuring mode 1 or mode 2 with writeMode method)? Is configuring these modes mandatory even if I want independent LED control? Is there a correlating value that I should specify with the independent LED hex address (default: 0000 0000) when setting them?

**Comment by ArtimusSteven** [comment]
I fixed the low output issue. Configuring mode2 to totem pole with other config bits outputs the correct 3.3V. Interesting because I was instructed to keep that bit as low drain.

**Comment by ArtimusSteven** [comment]
![delay_between_writes](https://user-images.githubusercontent.com/118404691/230681897-4655d0b1-f712-420a-a7fb-07c2b084c2ce.png)


This is the write delay of 270us i am referring to.

**Comment by RobTillaart** [comment]
In your code you use channel 15 and 16. These are adjacent numbers. 
 What happens if you use channel 14 and 16?
Does this change the 270 us? E.g. to 540 us?

**Comment by RobTillaart** [comment]
> > how did you determine this delay: On average one should see a delay of 1000000/380/2 = ~1300 us
> 
> The datasheet states the PWM uses a frequency of 97 KHz. So that is the freq of the individual pwm steps. These are rounded 10 us, so the PWM goes in 0..255 steps of ~10 us. Is after a max of ~2550 us the next PWM pulse starts over again.
> 
> As on average the register changes in the middle of such PWM cycle, it takes about half the cycle time before it becomes active. 2550/ 2 = ~1300 us.
> 
> That is the hypothesis.

Reread the datasheet again in greater detail.
Apparently the above hypothesis is wrong and not how it works. 
(It did not explain anything either so it was useless anyway)

The following image from the datasheet shows that the lowest level there are pulses of 40 nano seconds. 
256 of them makes 10.24 microseconds == 97.6 kHz.

![image](https://user-images.githubusercontent.com/462844/230716414-d2951eaa-609b-4557-8914-74a377308e21.png)

Did not find any mention that the 16 channels have a synchronized **t == 0** pulse. 
So it is not stated if they will PWM in sync or not, or if there is a fixed time between them.

----

A (good) reason to let the PWM's per channel start at a slightly different moment is that it would divide the load over the pins instead of a peak load at **t == 0**.  But again nothing found so far.

**Comment by RobTillaart** [comment]
Have you considered using the PCA9685? - https://www.nxp.com/docs/en/data-sheet/PCA9685.pdf

It allow you to adjust both the ON and OFF time per channel.

![image](https://user-images.githubusercontent.com/462844/230720715-d5ddac75-9188-41b9-87cd-c3cf88e36050.png)

From the datasheet 

#### 7.3.3 LED output and PWM control
_The turn-on time of each LED driver output and the duty cycle of PWM can be controlled 
independently using the LEDn_ON and LEDn_OFF registers.
There will be two 12-bit registers per LED output. These registers will be programmed by
the user. Both registers will hold a value from 0 to 4095. One 12-bit register will hold a
value for the ON time and the other 12-bit register will hold the value for the OFF time. The
ON and OFF times are compared with the value of a 12-bit counter that will be running
continuously from 0000h to 0FFFh (0 to 4095 decimal)._

Sounds like synchronous start per channel is well supported, and tunable.

**Comment by RobTillaart** [comment]
@ArtimusRobotics 
Did you make any progress, did any tests?

> In your code you use channel 15 and 16. 
These are adjacent numbers. 
What happens if you use channel 14 and 16? 
Does this change the 270 us? E.g. to 540 us?

Other tests done?

**Comment by ArtimusSteven** [comment]
Hey Rob, bringing focus back to this problem this week. Talked to another engineer working on the system and he stated that changing the pins to non adjacent numbers maintains the 270us delay. Next steps include adjusting the size of the pull up resistors for the SDA and SCL pins. Additionally, I am going to configure the chip for fast mode + to see if this improves delay as well. Not confident in changing the resistor values but could effect charge rate for delays of this size.

**Comment by ArtimusRobotics** [comment]
or fast mode+ **

**Comment by ArtimusSteven** [comment]
After further reading I believe the fast mode control is varied via those resistors.

**Comment by ArtimusSteven** [comment]
NEW FINDINGS: The delay between the 2 write commands decreases when the I2C clock on the teensy (SCL pin) is set to fast mode or fast mode plus. Lowering pull up resistor values and will report back.

**Comment by ArtimusSteven** [comment]
A recommendation: allow for users to define which speed mode they want to be in, in turn utilizing the Wire.setClock() method to change the speed of the I2C pin of the driving device (Standard, fast, fast plus).

**Comment by RobTillaart** [comment]
Hi, saw your comments and will read them tomorrow.
(busy with updating another library)

**Comment by RobTillaart** [comment]
Some reactions

> he stated that changing the pins to non adjacent numbers maintains the 270us delay.

Ok that implies the delay is not due to pin numbers. (does not help but is good to know)

----

> Next steps include adjusting the size of the pull up resistors for the SDA and SCL pins. 

Will only give a more reliable signal, but won't speed up things.

----

> Additionally, I am going to configure the chip for fast mode + to see if this improves delay as well. 

That is the interesting one, and yes, faster mode needs smaller pull ups to get a better signal.
See - https://www.gammon.com.au/forum/?id=10896&reply=5#reply5

**Comment by RobTillaart** [comment]
> NEW FINDINGS: The delay between the 2 write commands decreases when the I2C clock on the teensy (SCL pin) is set to fast mode or fast mode plus. Lowering pull up resistor values and will report back.

This is sort of expected as the I2C commands take time. And if communication is shorter the delay is shorter.

Q: Can you measure with a scope the time between the two commands to set the pins? 
You might toggle an "debug" IO line after the commands for easy measurement.

**Comment by RobTillaart** [comment]
> A recommendation: allow for users to define which speed mode they want to be in, in turn utilizing the Wire.setClock() method to change the speed of the I2C pin of the driving device (Standard, fast, fast plus).

They already have that freedom, however in practice the I2C bus speed is not set or set to the lowest speed supported by all. There are exceptions, but as long as the Wire Library does not support **getClockSpeed()** it is hard to optimize libraries. 
Setting speed to higher rates has also implications for pull ups and higher speeds give more trouble with e.g. in combination with longer wires. So only for those who understand I2C I would advice it.

**Comment by RobTillaart** [comment]
@ArtimusRobotics 

Went through the whole thread again, and the following idea popped up:

```cpp
loop(){
 curr = micros();

  if (sw1 && (curr - prev1) >= OnTime1)
  {
    uint8_t arr[2] = {0, 255};
    ledArray2.writeN(H15, arr, 2);    // <<<<<<<<<  
    sw1 = !sw1;
    prev1 = curr;
  }

  if ((curr - prev1) >= OffTime1)
  {
    uint8_t arr[2] = {255, 0};
    ledArray2.writeN(H15, arr, 2);    // <<<<<<<<<  
    sw1 = !sw1;
    prev1 = curr;
  }
}
```
As **write1()** is a wrapper around **writeN()** this should decrease the delay.
Both registers are updated in one I2C transaction.
Works only for adjacent pins...

**Comment by RobTillaart** [comment]
From the library
```cpp
/  write count values in consecutive PWM registers
//  checks if [channel + count - 1 > 15]
uint8_t PCA9635::writeN(uint8_t channel, uint8_t* arr, uint8_t count)
{
  if (channel + count > _channelCount)
  {
    _error = PCA9635_ERR_CHAN;
    return PCA9635_ERROR;
  }
  uint8_t base = PCA9635_PWM(channel);
  _wire->beginTransmission(_address);
  _wire->write(base);
  for(uint8_t i = 0; i < count; i++)
  {
    _wire->write(arr[i]);
  }
  _error = _wire->endTransmission();
  if (_error != 0)
  {
    _error = PCA9635_ERR_I2C;
    return PCA9635_ERROR;
  }
  _error = PCA9635_OK;
  return _error;
}
```

WriteN() does some error checking, removing this might make it a few usec faster.

**Comment by RobTillaart** [comment]
The library does not provide a way to use this register:

![image](https://user-images.githubusercontent.com/462844/234278431-29736b7d-da6b-4e96-ac62-7787336ad27c.png)

As it allows to write ON/OFF of 4 channels in one byte, so that could be a key to minimize the delay for adjacent channels.
(4 groups of 4 channels).
Again, the library has no such function (yet).

**Comment by RobTillaart** [comment]
@ArtimusRobotics 

Wrote a very simple wrapper around the LEDOUT registers.

You might add this to pca9635.h
```cpp
  //  writing reg 14-17  LEDOUT (7.3.6)
  //  group   leds
  //    0     0..3
  //    1     4..7
  //    2     8..11
  //    3     12..15
  uint8_t writeLedOut(uint8_t group, uint8_t mask);
  uint8_t readLedOut(uint8_t group);
```

You might add this to pca9635.cpp
```cpp
///////////////////////////////////////////////////////////////
//
//  EXPERIMENTAL LEDOUT
//
uint8_t PCA9635::writeLedOut(uint8_t group, uint8_t mask)
{
  if (group > 3) return PCA9635_ERROR;
  writeReg(PCA9635_LEDOUT_BASE + group, mask);
  return PCA9635_OK;
}

uint8_t PCA9635::readLedOut(uint8_t group)
{
  if (group > 3) return 0x00;
  return readReg(PCA9635_LEDOUT_BASE + group);
}
```

----
updated code

**Comment by ArtimusSteven** [comment]
> There are exceptions, but as long as the Wire Library does not support **getClockSpeed()** it is hard to optimize libraries.

I wonder, if there are any, what are challenges to implementing a getClockSpeed method? I'd think the speed would be stored in some register on the Teensy (i am using a 3.6). I'll take a further look at the Wire library.

**Comment by RobTillaart** [comment]
> I wonder, if there are any, what are challenges to implementing a getClockSpeed() method?

- technically it is not that difficult. 
  - 1st order implementation could just cache the parameter. 
  - 2nd order is to reverse the calculations (and possibly get a slightly deviating value)
- it must be supported by all boards. otherwise you can't use the call safely in a library. 
  - it won't compile if a board does not support it. 
  - https://github.com/arduino/Arduino/issues/11457 - open for 2 yrs now.

imho the biggest problem is point 2.

Having a getClockSpeed() one can 
- ask the current speed, 
- store it. 
- Set a device max speed,  (configurable in the API)
- do your action, 
- restore the previous speed.

That way every device could be operated at max speed feasible.

(from some earlier tests)

|  proc  |  support |
|:-------:|:---------:|
|  ESP32 | Y |
|  ESP8266 | N |
| AVR UNO |  N  |
| Nano 33 | N |
|  MKR1000 | N |
| Teensy | N |

As far as I know there is no compile time test to see if a function is supported or not (like #ifdef).

To get it supported by Teensy => contact Paul Stoffregen, he is open for good ideas.

**Comment by ArtimusSteven** [comment]
ESP32 | Y
ESP8266 | N
AVR UNO | N
Nano 33 | N
MKR1000 | N
Teensy | N

What does it look like to ensure a method like getClock() can run on all chips listed? ^^
Create if cases that check for board defined? Obviously need to understand I2C registers on all chips.

uint32_t  TwoWire::getClock()
{
 uint32_t speed = F_CPU / ((TWBR * 2) + 16);
 return speed;
}

this is the method you wrote. This would need to defined across all boards. So at the very least identifying the F_CPU and TWBR across all chips. Could you explain the math quickly to get the speed?

**Comment by ArtimusSteven** [comment]
Also, I believe your new writeN suggestion reduced the delay down to nano seconds. I will include oscilloscope shots shortly.

**Comment by RobTillaart** [comment]
> What does it look like to ensure a method like getClock() can run on all chips listed? ^^
this is the method you wrote. This would need to defined across all boards. So at the very least identifying the F_CPU and TWBR across all chips. Could you explain the math quickly to get the speed?

F_CPU is a #define that is supported by most boards (core SW) it is the CPU clock. The magnitude determines the max rate of the I2C bus. An UNO can do 800KHz max, A Teensy probably 3.4 MHz.
Some boards have multiple possible clockspeeds e.g ESP32 which makes things "more diverse/complex?"

TWBR is AVR specific (think it stands for TwoWireBit-Rate or TwoWireBit-Register, other boards use different registers etc.
So other boards have different formulas and registers. 

All manufactures (or at least a substantial part) should agree to implement to get it mainstream. 

----

wrt the code 

```cpp
uint32_t TwoWire::getClock()
{
uint32_t speed = F_CPU / ((TWBR * 2) + 16);
return speed;
}
```
The **Wire.setClock()** call for  AVR chips Atmel 328 et al use the F_CPU (frequency CPU) and the wanted clock speed to set the TWBR register. There is optionally another register involved, a divider but I do not know it from my head. It is seldom used, only for very low bit rates. 
The code above is the "reverse" of the Wire.setClock() code, and as the setClock() has truncating in it, values are not exact (but indicative enough. This is from core of AVR Wire 
```cpp
void twi_setFrequency(uint32_t frequency)
{
  TWBR = ((F_CPU / frequency) - 16) / 2;
}
```

**Comment by RobTillaart** [comment]
> Also, I believe your new writeN suggestion reduced the delay down to nano seconds. I will include oscilloscope shots shortly.

I'm interested to hear/see the results

**Comment by ArtimusSteven** [comment]
It appears that a single writeN to all 16 channels while focusing on updating the individual "pin values" in the array allow for the fastest individual LED control for the chip (~100ns rise and fall times). I tried also writing to individual channels with the writeN instead of the write 1; but the delay increased to ~130us. 

![scope_0](https://user-images.githubusercontent.com/118404691/234392394-c628cf6a-ba33-4c58-a95b-d9d909665158.png)

here is a scope snippet of the ~100ns delay between a low write of pin 12 and a high write of pin 0. This confirms that thus far a single call to written with updates to array indexes is fastest option for individual control.

**Comment by ArtimusSteven** [comment]
> > What does it look like to ensure a method like getClock() can run on all chips listed? ^^
> > this is the method you wrote. This would need to defined across all boards. So at the very least identifying the F_CPU and TWBR across all chips. Could you explain the math quickly to get the speed?
> 
> F_CPU is a #define that is supported by most boards (core SW) it is the CPU clock. The magnitude determines the max rate of the I2C bus. An UNO can do 800KHz max, A Teensy probably 3.4 MHz. Some boards have multiple possible clockspeeds e.g ESP32 which makes things "more diverse/complex?"
> 
> TWBR is AVR specific (think it stands for TwoWireBit-Rate or TwoWireBit-Register, other boards use different registers etc. So other boards have different formulas and registers.
> 
> All manufactures (or at least a substantial part) should agree to implement to get it mainstream.
> 
> wrt the code
> 
> ```c++
> uint32_t TwoWire::getClock()
> {
> uint32_t speed = F_CPU / ((TWBR * 2) + 16);
> return speed;
> }
> ```
> 
> The **Wire.setClock()** call for AVR chips Atmel 328 et al use the F_CPU (frequency CPU) and the wanted clock speed to set the TWBR register. There is optionally another register involved, a divider but I do not know it from my head. It is seldom used, only for very low bit rates. The code above is the "reverse" of the Wire.setClock() code, and as the setClock() has truncating in it, values are not exact (but indicative enough. This is from core of AVR Wire
> 
> ```c++
> void twi_setFrequency(uint32_t frequency)
> {
>   TWBR = ((F_CPU / frequency) - 16) / 2;
> }
> ```

What are the roles of the values 2 and 16?

**Comment by RobTillaart** [comment]
> It appears that a single writeN to all 16 channels while focusing on updating the individual "pin values" in the array allow for the fastest individual LED control for the chip (~100ns rise and fall times).

Is this in the range of what you need?
Image looks very good imho. You can make the edges steeper by smaller resistors.

Did you find time to test the LEDOUT functions? These can write 4 channels with one byte.

Snippet
```cpp
{
  uint8_t mask = readLedOut(3);
  mask &= 0b00001111;    //  LED 14 + LED15 OFF
  mask |= 0b00010000;     //  LED14 ON
  writeLedOut(3, mask);
}
else
{
  uint8_t mask = readLedOut(3);
  mask &= 0b00001111;    //  LED 14 + LED15 OFF
  mask |= 0b01000000;     //  LED15 ON
  writeLedOut(3, mask);
}
....

**Comment by RobTillaart** [comment]
> This is from core of AVR Wire
```cpp
void twi_setFrequency(uint32_t frequency)
{
  TWBR = ((F_CPU / frequency) - 16) / 2;
}
```
> What are the roles of the values 2 and 16?

Did some googles and found that in the AVR chip the SCL frequency should at least be 16 times lower than the FCPU.
There was no reason explained why. I assume to give the hardware time to get a byte and shift it into bits and send the bits.

https://onlinedocs.microchip.com/pr/GUID-EF6FA2F3-5345-4829-9F5D-75388E2EF0C9-en-US-4/index.html?GUID-3BBC8619-26AF-4885-BB0C-D234E30426BD

The factor 2 is probably (quite sure) used to keep the TWBR into a single byte while keeping a pragmatic operational range.
TWBR == 0 ==>  SCL = 1MHz
TWBR == 255  ==> SCL = 30.4 KHz.

Maybe more info is in the datasheet.

**Comment by RobTillaart** [comment]
https://www.electronicwings.com/avr-atmega/atmega1632-i2c  explains even a bit more (Assume 328 behaves similar)

The formula in the link above looks incorrect, the - (minus) should be * (times)

https://www.gammon.com.au/i2c is also a good source about I2C on the UNO.

**Comment by RobTillaart** [comment]
Created a develop branch + PR (0.4.5) to include the read- and writeLedOut() registers.

**Comment by ArtimusSteven** [comment]
> @ArtimusRobotics
> 
> Wrote a very simple wrapper around the LEDOUT registers.
> 
> You might add this to pca9635.h
> 
> ```c++
>   //  writing reg 14-17  LEDOUT (7.3.6)
>   //  group   leds
>   //    0     0..3
>   //    1     4..7
>   //    2     8..11
>   //    3     12..15
>   uint8_t writeLedOut(uint8_t group, uint8_t mask);
>   uint8_t readLedOut(uint8_t group);
> ```
> 
> You might add this to pca9635.cpp
> 
> ```c++
> ///////////////////////////////////////////////////////////////
> //
> //  EXPERIMENTAL LEDOUT
> //
> uint8_t PCA9635::writeLedOut(uint8_t group, uint8_t mask)
> {
>   if (group > 3) return PCA9635_ERROR;
>   writeReg(PCA9635_LEDOUT_BASE + group, mask);
>   return PCA9635_OK;
> }
> 
> uint8_t PCA9635::readLedOut(uint8_t group)
> {
>   if (group > 3) return 0x00;
>   return readReg(PCA9635_LEDOUT_BASE + group);
> }
> ```
> 
> updated code

I will try these from the develop branch and report back

**Comment by RobTillaart** [comment]
If you have ideas for additional generic functionality, let me know.

I am thinking of **allOff()** and **allOn()**.

Furthermore a **shiftUp()** and **shiftDown()**. These read all 16 channels move them one channel, and add one new value. This might be just an example not perse in the library.
In the same category one could have **rotate**. Thes all come down to manipulating a 16 byte array and call **writeN()**

**Comment by RobTillaart** [comment]
- added **setLedDriverMode(uint8_t mode)** to set all channels in one go. (allON, allOFF, allPWM).
- added example to show how to shift and rotate channels in a simple way.

build is running

**Comment by RobTillaart** [comment]
@ArtimusRobotics 

> I will try these from the develop branch and report back

Any problems found? 
If not I will merge current develop into master. 
(latest push was a sync with PCA9634 + redo of constants to bring these two libs even closer.)

**Comment by RobTillaart** [comment]
Syncing with PCA9634 took a bit more time than expected. To be continued tomorrow.

**Comment by ArtimusSteven** [comment]
Could you explain how the mask works for the new writeLedOut method? I understand it is a more efficient way to control 4 leds with one byte of information; but where does the actual pwm value(0-255) for each pin apply?

**Comment by RobTillaart** [comment]
There are 4 registers
Every register is 8 bits, 4 pairs of 2.
Each pair indicates the configuration of one channel.
The options are (from my head)
- 00 LED OFF
- 01 LED ON
- 10 PWM mode - uses 1 pwm register per channel
- 11 PWMGRP mode - see datasheet.

To change the mode of a channel, you must manipulate the right pair of bits. That typically includes
- read the current register
- clear the bits you want to set ==> and operation
- set the bits you want to == > or operation
- write back the register.

To get the pair of bits in the right spot you need to shift them to the left.

That is it in a nutshell,

**Comment by ArtimusSteven** [comment]
Hey Rob, could you explain the bit magic you are doing to set the led driver mode? Specifically the logic here: 

uint8_t PCA9635::setLedDriverMode(uint8_t channel, uint8_t mode)
{
  if (channel >= _channelCount)
  {
    _error  = PCA963X_ERR_CHAN;
    return _error;
  }
  if (mode > 3)
  {
    _error  = PCA963X_ERR_MODE;
    return _error;
  }

  uint8_t reg = PCA963X_LEDOUT_BASE + (channel >> 2);
  //  some bit magic
  uint8_t shift = (channel & 0x03) * 2;  // 0,2,4,6 places
  uint8_t setmask = mode << shift;
  uint8_t clrmask = ~(0x03 << shift);
  uint8_t value = (readReg(reg) & clrmask) | setmask;
  writeReg(reg, value);
  _error = PCA963X_OK;
  return _error;
}


//  returns 0..3 if OK, other values indicate an error
uint8_t PCA9635::getLedDriverMode(uint8_t channel)
{
  if (channel >= _channelCount)
  {
    _error  = PCA963X_ERR_CHAN;
    return _error;
  }

  uint8_t reg = PCA963X_LEDOUT_BASE + (channel >> 2);
  uint8_t shift = (channel & 0x03) * 2;  //  0, 2, 4, 6 places
  uint8_t value = (readReg(reg) >> shift ) & 0x03;
  _error = PCA963X_OK;
  return value;
}

**Comment by RobTillaart** [comment]
SetLedDriveMode
It reads the register
Clears the bits on the position of the channel
Shifts the requested mode bits to match that position
Fill those bits in the mask
Writes the mask back.

The reading is a subset of the above.

You might need to read a tutorial about bit operations in C. I expect these explain this techniques far better than I can.

If time permits I will add comments to the code before merging the PR.

**Comment by RobTillaart** [comment]
I merged the PR and released the 0.4.5 version.
If there are no further questions please close the issue.

**Comment by ArtimusSteven** [comment]
Do you have any ideas regarding how to enforce the timing of pwm signals based off of a serial communication script that sends commands to a teensy script in order to send commands to the LED Driver? 

As of now, i am beginning to understand the delay experienced. When performing two consecutive pwm register write procedures (using the wire library), there is a 40us delay between these two writes. When the writes take advantage of the auto increment feature on the chip, this delay increases further. My issue currently is with enforcing the timing at which the pwm pins are communicated to. 


    // curr = micros();
    if (Serial.available() > 0)
    {

        // static int message_read_flag = 0;

        static char message[MAX_MESSAGE_LENGTH];
        static unsigned int message_pos = 0;
        char inByte = Serial.read();

        if (inByte != '\n' && (message_pos < MAX_MESSAGE_LENGTH - 1))
        {
            // Add the coming byte to our message
            message[message_pos] = inByte;
            message_pos++;
        }

        else
        {

            // if (sw1 && (curr - prev1) >= OnTime1)
            if (message[0] == '1')
            {
                message_pos = 0;
                // uint8_t x1 = 0;
                // uint8_t x2 = 255;
                // uint8_t arr[16] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
                // arr[H15] = 255;
                // arr[H13] = 0;
                if (actuate1 == 1)
                {
                    Wire.beginTransmission(chip59);
                    Wire.write(0x02);
                    Wire.write(0xFF);
                    Wire.endTransmission(true);
                    actuate1 = 2;
                }
                else if (actuate1 == 2)
                {
                    Wire.beginTransmission(chip59);
                    Wire.write(0x02);
                    Wire.write(0x00);
                    Wire.endTransmission(true);
                    actuate1 = 1;
                }
            }

            if (message[0] == '2')
            {
                message_pos = 0;
                // uint8_t x1 = 0;
                // uint8_t x2 = 255;
                // uint8_t arr[16] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
                // arr[H15] = 0;
                // arr[H13] = 255;
                // ledArraactuate2.writeN(H13, arr, 16
                if (actuate2 == 1)
                {
                    Wire.beginTransmission(chip59);
                    Wire.write(0x0E);
                    Wire.write(0xFF);
                    Wire.endTransmission(true);
                    actuate2 = 2;
                }
                else if (actuate2 == 2)
                {
                    Wire.beginTransmission(chip59);
                    Wire.write(0x0E);
                    Wire.write(0x00);
                    Wire.endTransmission(true);
                    actuate2 = 1;
                }
            }
        }
    }

This is some test code i am working with. There is another serial script that sends the actuate commands. The delays experienced here begin to compound. Ideally these two different pwm registers are written to every 1/2 and 1 Hz and track together. Instead they are written to every 0.5 and 1 second, but the timing of the two together start to drift apart.

**Comment by RobTillaart** [comment]
> Do you have any ideas regarding how to enforce the timing of pwm signals based off of a serial communication script that sends commands to a teensy script in order to send commands to the LED Driver?

Maybe you should reason backwards, in order to send commands in time to the LED driver, the teensy should have them BEFORE they need to be send to the PCA. The teensy can keep track of timing in micros() so drift can be kept minimal.

So the idea is to let the teensy hold a (small) queue per channel and let the teensy control the timing. 
This means that the commands should be send over serial a few 100 millis in advance.

Don't know if that is possible in your project ?

**Comment by ArtimusSteven** [comment]
This protocol is something i will begin to develop for the project i am working on. Drift is my biggest issue right now. I will keep you updated.

**Comment by ArtimusSteven** [comment]
If i close this issue, will I be able to refer back to it in the future as a resource?

**Comment by RobTillaart** [comment]
Yes, and tou can still comment and even reopen it if needed. (New topic == new issue).

I never delete issues as they often contain a lit of rationale why something changed, sometimes alternatives considered etc. In short valuable.

**Comment by ArtimusSteven** [comment]
Hey Rob, I've been continuing to work on this issue. I have been able to reduce delay between led channel pin writes via an increase in the I2C clock to 1MgHz and logic that writes to pins consecutively either in groups of 4 or 16. I've attempted a type of buffer that reads for 500 us and then writes to the channels in a loop manner. While I am not positive this solution is helping, the approach aims to collect pin write commands within a small enough window in order to write to multiple pins within one iteration of the control  loop. 
In regards to drift, it is still very much present with this new setup. I wanted to ask what you know about the concept of drift itself with these signals. Where and why is drift occurring in a system that sends write commands over a serial port based on defined frequencies, receives those commands via a teensy serial.read loop that reads for 500us intervals, and writes to pins after each buffered reading cycle? Any thoughts you have on this concept to prompt progress would be greatly appreciated.

**Comment by RobTillaart** [comment]
Good to hear that 1MHz working as faster I2C does decrease the delay.

**Drift** 
My most important understanding of digital electronics is that on lowest level it is analog and not exact.

The accuracy of digital only goes to some extend. The frequencies of the serial port are accurate enough to handle frequencies +- few percent.  e.g. a port set on 9600 baud will probably work with a signal that is between 9500 and 9700. A small deviation of only 1 bps will add up over time.

Another cause for drift is how I2C incoming data is handled. Typical it is clocked (with its own accuracy) into a buffer and an interrupt is generated, which is handled as soon as possible, implying also variation in timing as there might be other tasks with higher priority.

A digital signal is HIGH or LOW, but these are defined as e.g. 0-2V and 3-5V, there is a gray area where a flipping line is not defined. Flipping a line from LOW to HIGH or vice versa takes time (RC curve), a small period but also with its own variation as the RC depends on wires connected, the capacitance and resistance.

In short all components drift due to the analog nature of electronics and its components. 

----

If you have set the timing for the 16 pins, they start at a different moment. (problem in short).
Does this "delta" stay in every cycle or does the start of the pulses become synchronized e.g. after one or two pulses?

If they disappear after a cycle, you might add an AND port after each port and enable them (all) with one single line.

```
               PCA9635                AND
             +----------+            +---+
             |       0  |----------->|   |
             |          |            | & |----->0'
             |          |       +--->|   |
             |          |       |    +---+
             |          |       |
             |          |       |
             |          |       |    +---+
             |       1  |----------->|   |
             |          |       |    | & |----->1'
             |          |       +--->|   |
             |          |       |    +---+
             |          |       |
             |          |       |
             |          |       |    +---+
             |       2  |----------->|   |
             |          |       |    | & |----->2'
             |          |       +--->|   |
             |          |       |    +---+
             |          |       |
             |   16 x   |       |
             |          |       |
             |          |       |
             |          |       |
             |          |       |
             |          |       |
             +----------+       |
                                |
                                |
                              ENABLE
```

you set pin 0 , 1 , 2 etc as fast as possible, but only after **ENABLE** gets HIGH all outputs 0', 1', 2' will follow the PCA9635.

Don't know if such setup fits into your project?


(you can adapt this to the subset of lines you need, so more flexible than the OE line of the PCA chip)

**Comment by RobTillaart** [comment]
Back to the start.
The system needs to control 32 LED lights. What is the purpose of those 32 LEDs? 
Why is this timing so important?

----

To get maximum speed, you might need to let the I2C PCA9635 go and switch to a SPI based look-alike TLC5947.
- https://learn.adafruit.com/tlc5947-tlc59711-pwm-led-driver-breakout/overview

The TLC5947 (== 24 channels of 12 bit) can handle 15 MHz in cascaded mode and 30 MHz standalone.
This is roughly a factor 10 above your current I2C device speed.
This performance gain might compensate for the fact that you need to send more data (24x12 bit)

you can use 2 TLC5947 in cascaded mode or side by side so you can control 2x 16 pins (at 12 bit resolution)
You might want to use 4 of these side by side and only use 8 ports per TLC to optimize performance. 

I never worked with them (don't have a library for them either)

**Comment by RobTillaart** [comment]
@ArtimusRobotics 
Dived into the datasheet of the TLC5947 it looks more than interesting.  24 channels with 12 bit PWM.
Other important feature is it has a latch pin to all 24 channels, which start in sync in 4 groups of 6.


![image](https://github.com/RobTillaart/PCA9635/assets/462844/a72c3064-5d09-434c-b641-e7d2eed47480)

![image](https://github.com/RobTillaart/PCA9635/assets/462844/c056f3d7-115e-46b9-b745-23690710c6e6)

So max 0.1 usec delta between start of pulses.

Started writing a library, experimental as I have no hardware, initial version in develop branch
- https://github.com/RobTillaart/TLC5947

**Comment by RobTillaart** [comment]
FYI
Released version 0.1.0 - https://github.com/RobTillaart/TLC5947

**Comment by RobTillaart** [comment]
- https://github.com/RobTillaart/TLC5947

has made its first release 2 days ago, expect a performance update soon.
